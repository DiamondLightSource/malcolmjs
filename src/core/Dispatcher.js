/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
