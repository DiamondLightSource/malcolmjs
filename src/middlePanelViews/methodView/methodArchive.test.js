import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import MockCircularBuffer from '../../malcolm/reducer/attribute.reducer.mocks';

import MethodArchive from './methodArchive.container';

describe('MethodArchive', () => {
  let shallow;
  let testArchive;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
    testArchive = {
      value: new MockCircularBuffer(5),
      timeStamp: new MockCircularBuffer(5),
      alarmState: new MockCircularBuffer(5),
    };
    testArchive.value.push({
      runParameters: { test1: 0 },
      returned: { test2: 1 },
    });
    testArchive.timeStamp.push({
      localRunTime: new Date(0),
      localReturnTime: new Date(),
    });
    testArchive.alarmState.push(0);
  });

  it('renders table correctly for input param', () => {
    const wrapper = shallow(
      <MethodArchive
        methodArchive={testArchive}
        openPanels={{}}
        selectedParam={['takes', 'test1']}
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders table correctly for return param', () => {
    const wrapper = shallow(
      <MethodArchive
        methodArchive={testArchive}
        openPanels={{}}
        selectedParam={['returns', 'test2']}
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders 2nd tab correctly', () => {
    const wrapper = shallow(
      <MethodArchive
        methodArchive={testArchive}
        openPanels={{}}
        defaultTab={1}
        selectedParam={['takes', 'test1']}
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
