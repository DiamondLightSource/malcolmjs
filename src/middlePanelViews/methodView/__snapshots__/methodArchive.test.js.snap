// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MethodArchive renders 2nd tab correctly 1`] = `
<Connect(WithTheme(TabbedPanel))
  alwaysUpdate={true}
  defaultTab={1}
  openPanels={Object {}}
  tabLabels={
    Array [
      "Table",
      "Plot",
    ]
  }
>
  <WithStyles(WidgetTable)
    addRow={[Function]}
    attribute={
      Object {
        "calculated": Object {},
        "raw": Object {
          "meta": Object {
            "elements": Object {
              "alarm": Object {
                "label": "Method Alarm state",
                "tags": Array [
                  "info:alarm",
                ],
              },
              "timeStamp": Object {
                "label": "Time sent",
                "tags": Array [
                  "widget:textupdate",
                ],
              },
              "value": Object {
                "label": "Parameter value",
                "tags": Array [
                  "widget:textupdate",
                ],
              },
            },
          },
          "value": Object {
            "alarm": Array [
              0,
            ],
            "timeStamp": Array [
              "1970-01-01T00:00:00.000Z",
            ],
            "value": Array [
              0,
            ],
          },
        },
      }
    }
    closePanelHandler={[Function]}
    eventHandler={[Function]}
    hideInfo={true}
    infoClickHandler={[Function]}
    rowClickHandler={[Function]}
    setFlag={[Function]}
  />
  <LoadableComponent
    attribute={
      Object {
        "alarmState": Array [
          0,
        ],
        "name": undefined,
        "parent": undefined,
        "timeStamp": Array [
          1970-01-01T00:00:00.000Z,
        ],
        "value": Array [
          0,
        ],
      }
    }
    deriveState={[Function]}
    openPanels={Object {}}
  />
</Connect(WithTheme(TabbedPanel))>
`;

exports[`MethodArchive renders table correctly for input param 1`] = `
<Connect(WithTheme(TabbedPanel))
  alwaysUpdate={true}
  defaultTab={1}
  openPanels={Object {}}
  tabLabels={
    Array [
      "Table",
      "Plot",
    ]
  }
>
  <WithStyles(WidgetTable)
    addRow={[Function]}
    attribute={
      Object {
        "calculated": Object {},
        "raw": Object {
          "meta": Object {
            "elements": Object {
              "alarm": Object {
                "label": "Method Alarm state",
                "tags": Array [
                  "info:alarm",
                ],
              },
              "timeStamp": Object {
                "label": "Time sent",
                "tags": Array [
                  "widget:textupdate",
                ],
              },
              "value": Object {
                "label": "Parameter value",
                "tags": Array [
                  "widget:textupdate",
                ],
              },
            },
          },
          "value": Object {
            "alarm": Array [
              0,
            ],
            "timeStamp": Array [
              "1970-01-01T00:00:00.000Z",
            ],
            "value": Array [
              0,
            ],
          },
        },
      }
    }
    closePanelHandler={[Function]}
    eventHandler={[Function]}
    hideInfo={true}
    infoClickHandler={[Function]}
    rowClickHandler={[Function]}
    setFlag={[Function]}
  />
  <LoadableComponent
    attribute={
      Object {
        "alarmState": Array [
          0,
        ],
        "name": undefined,
        "parent": undefined,
        "timeStamp": Array [
          1970-01-01T00:00:00.000Z,
        ],
        "value": Array [
          0,
        ],
      }
    }
    deriveState={[Function]}
    openPanels={Object {}}
  />
</Connect(WithTheme(TabbedPanel))>
`;

exports[`MethodArchive renders table correctly for return param 1`] = `
<Connect(WithTheme(TabbedPanel))
  alwaysUpdate={true}
  defaultTab={1}
  openPanels={Object {}}
  tabLabels={
    Array [
      "Table",
      "Plot",
    ]
  }
>
  <WithStyles(WidgetTable)
    addRow={[Function]}
    attribute={
      Object {
        "calculated": Object {},
        "raw": Object {
          "meta": Object {
            "elements": Object {
              "alarm": Object {
                "label": "Method Alarm state",
                "tags": Array [
                  "info:alarm",
                ],
              },
              "timeStamp": Object {
                "label": "Time received",
                "tags": Array [
                  "widget:textupdate",
                ],
              },
              "value": Object {
                "label": "Parameter value",
                "tags": Array [
                  "widget:textupdate",
                ],
              },
            },
          },
          "value": Object {
            "alarm": Array [
              0,
            ],
            "timeStamp": Array [
              "1970-01-01T00:00:00.000Z",
            ],
            "value": Array [
              undefined,
            ],
          },
        },
      }
    }
    closePanelHandler={[Function]}
    eventHandler={[Function]}
    hideInfo={true}
    infoClickHandler={[Function]}
    rowClickHandler={[Function]}
    setFlag={[Function]}
  />
  <LoadableComponent
    attribute={
      Object {
        "alarmState": Array [
          0,
        ],
        "name": undefined,
        "parent": undefined,
        "timeStamp": Array [
          1970-01-01T00:00:01.000Z,
        ],
        "value": Array [
          undefined,
        ],
      }
    }
    deriveState={[Function]}
    openPanels={Object {}}
  />
</Connect(WithTheme(TabbedPanel))>
`;
